core: basic UI support for T1 (v3)

mmilata opened the pull request:
Obsoletes #1239. Introduces the _widget_ abstraction, in the form of model-specific callable (async) functions. Drops the requirement that only strings and lists be passed as layout/widget parameters, since it's not clear whether we actually want to send these in ButtonRequests.

Notes:
* No progress handling yet (waiting for input).
* Wire context is passed into widgets, as well as ButtonRequest fields (where these are not hardcoded into widget), looking into whether it makes sense to treat these as single object that will be passed to the widgets.
* Different wording for different models will be handled by the translation logic which doesn't exist yet so there's currently some TT texts on T1.
![emu00000000](https://user-images.githubusercontent.com/85857/96193025-8a297700-0f47-11eb-9017-96edd0843ef2.png)

andrewkozlik commented:
I have just had a brief look at this PR and conceptually I like it! It seems to me that this is the correct division of responsibilities in the codebase. I will look at the details a bit more later, but I leave the rigorous code review to someone else. I have two comments right now:
- As I mentioned in the design doc, I wouldn't call functions like `confirm_change_count_over_threshold()` "widgets", because to me widgets are buttons, labels, text boxes etc. So something else like "dialogs". Anyone have a better name?
- It would be good to implement two line breaking policies, which would probably be an additional parameter of `render_text()`:
    - The one we have now with the dashes, which is useful for displaying long strings like addresses.
    - One which breaks lines only at spaces, so it looks ahead to see if the next word fits on the line. This would save us the burden of manually having to figure out where to add '\n' at different places for the T1 and TT. See the screenshot of "Do you want to create a new wallet?" above.

matejcik commented on core/src/trezor/ui/layouts/tt.py:
```diff
@@ -0,0 +1,293 @@
+    is_multisig = len(xpubs) > 0
+    while True:
+        if (
+            await interact(
```
Currently `show_address` does not send intermediate button requests. I think we intended to keep that behavior?

mmilata replied:
Actually it does - in master `show_address` as well as `show_qr` send BR and the application code calls them from a while loop. So the current behavior is kept afaict.

matejcik reviewed (commented):
I pushed a bunch of features.

Namely:

- `Text.format_parametrized`, which does two important things:
  * automatically breaks texts on `\n`
  * can insert a bolded parameter denoted by `{}` in the string
  both will be useful for translations, where I expect the localization string will look like `"The fee of {}\nis too high"`, or `"Transakční poplatek\nve výši {}\nje příliš velký"`
- reduced specificity of some widgets. There is now a general `confirm_metadata` for anything (so far) that is not output or total, `confirm_hex` for any sort of data blob (will be useful for Ethereum data, presumably, and maybe some key material?)
  I believe `confirm_wipe` should also be generalized, or maybe even folded into `confirm_action`. I would keep `confirm_setup` (it's distinct by the TOS thing), and ideally get rid of `confirm_backup` subscreens. But I deferred these until we see the UI redesign.
  I don't like `confirm_joint_total` separate from `confirm_total`, but I don't see a good way to generalize it either. So let's keep it separate for now.
- renamed `widgets` to `layouts` and changed the package layout, so that `trezor/ui/layouts` and `trezor/ui/components` are separate. This just seems better for the moment when the widgets grow too large to keep in the same file. App code should ideally avoid `trezor.ui.components` completely in the future.
  Admittedly it's a little awkward to have to import `from ..components.tt` in layouts, but it's not that bad, it seems?
- dropped pixel-perfectness of some screens, in favor of better layout function design

couple minor points in the review comments

high-level:

* `type` should contain the word "confirm" much less than it does. I removed it from the metadata dialogs, but it remained somewhere.
  Or we should ask Suite if some naming convention would make it easier for them.
* the more generic layouts (currently only `confirm_action`) will need a `style` argument, instead of `icon`. There should be a very small library of styles -- we need to get this from the UI designer!
  The advantage of "style" over "icon" is that a style can include colors, button sizes, etc. These might apply on TT but not on T1.
* I didn't implement it in the end, but we might need to have a "top text" and "bottom text". It seems that instead of the `Text` class carrying header, `Confirm` (and possibly other layout classes) should have it, then a top-aligned block of text and then a bottom (button) aligned block of text. This way we get rid of various `br_half`s and achieve a more consistent look. Again, this depends a lot on what the UI design brings.

---

overall, I think this is going in the right direction, and that the next step is moving more application code to use layouts

andrewkozlik commented:
I haven't looked at the code, but the new features sound awesome.

> Text.format_parametrized, which does two important things:
> * automatically breaks texts on \n
> * can insert a bolded parameter denoted by {} in the string
	
I am wondering if it wouldn't be better to use `\1`, `\2`, ..., `\9` instead of `{}`. When there are multiple parameters some languages may need to permute them due to their specific subject/object/verb word order. We can always implement the python formatting syntax like `{:1}` or `{amount}`, but `\1` would probably be simpler to implement.

Also I am not entirely sure that the automatic bolding is always desirable, but we can defer this until after the UI redesign.

matejcik commented:
> I am wondering if it wouldn't be better to use `\1`, `\2`, ..., `\9` instead of `{}`

good point, we can add that in later

> automatic bolding is always desirable

if you don't want automatic bolding, you can insert the parameter via normal string formatting, before passing into the `Text` class :)

mmilata commented on core/src/trezor/ui/components/common/text.py:
```diff
@@ -0,0 +1,180 @@
+    TEXT_HEADER_HEIGHT = None  # type: int
+    TEXT_LINE_HEIGHT = None  # type: int
+    TEXT_LINE_HEIGHT_HALF = None  # type: int
+    TEXT_MARGIN_LEFT = None  # type: int
```
Fixed in 5be66d5546f7243d5c0ebd731e4c1cf149452959.

mmilata commented on core/src/trezor/ui/components/t1/text.py:
```diff
@@ -0,0 +1,58 @@
+
+    if __debug__:
+
+        def read_content(self) -> List[str]:
```
Fixed in 5be66d5546f7243d5c0ebd731e4c1cf149452959.

mmilata commented:
Rebased onto `master` and added couple of fixes, mostly to see CI results. No need to review, I'm later going to split the changes into three branches:
1. refactoring that does not affect behavior or UI look, to be merged soon
2. branch to work on TT redesign, to be merged after all screens are redone
3. T1 support, to be merged some time after 2.
